fuse_viz/include/fuse_viz/mapped_covariance_property.h

```diff
@@ -0,0 +1,129 @@
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef MAPPED_COVARIANCE_PROPERTY_H
+#define MAPPED_COVARIANCE_PROPERTY_H
+
+#include <QColor>
+
+#include <OgreColourValue.h>
+
+#include <rviz/properties/bool_property.h>
+
+#include <string>
+#include <unordered_map>
+
+namespace Ogre
+{
+
+class SceneManager;
+class SceneNode;
+
+}  // namespace Ogre
+
+namespace rviz
+{
+
+class Property;
+class ColorProperty;
+class FloatProperty;
+class EnumProperty;
+class MappedCovarianceVisual;
+
+/**
+ * @brief Property specialized to provide getter for booleans.
+ *
+ * This is mostly a copy of CovarianceProperty from rviz/default_plugin/covariance_property.h that instead of using an
+ * std::deque to store the visuals, it uses an std::unordered_map, so the visuals can be indexed by their UUID.
+ */
+class MappedCovarianceProperty : public rviz::BoolProperty
+{
+  Q_OBJECT
+public:
+  typedef boost::shared_ptr<MappedCovarianceVisual> MappedCovarianceVisualPtr;
+
+  enum Frame
+  {
+    Local,
+    Fixed,
+  };
+
+  enum ColorStyle
+  {
+    Unique,
+    RGB,
+  };
+
+  MappedCovarianceProperty(const QString& name = "Covariance", bool default_value = false,
+                           const QString& description = QString(), rviz::Property* parent = 0,
+                           const char* changed_slot = 0, QObject* receiver = 0);
+
+  virtual ~MappedCovarianceProperty();
+
+  bool getPositionBool();
+  bool getOrientationBool();
+
+  // Methods to manage the unordered map of Covariance Visuals
+  MappedCovarianceVisualPtr createAndInsertVisual(const std::string& key, Ogre::SceneManager* scene_manager,
+                                                  Ogre::SceneNode* parent_node);
+  void eraseVisual(const std::string& key);
+  void clearVisual();
+  size_t sizeVisual();
+
+public Q_SLOTS:
+  void updateVisibility();
+
+private Q_SLOTS:
+  void updateColorAndAlphaAndScaleAndOffset();
+  void updateOrientationFrame();
+  void updateColorStyleChoice();
+
+private:
+  void updateColorAndAlphaAndScaleAndOffset(const MappedCovarianceVisualPtr& visual);
+  void updateOrientationFrame(const MappedCovarianceVisualPtr& visual);
+  void updateVisibility(const MappedCovarianceVisualPtr& visual);
+
+  std::unordered_map<std::string, MappedCovarianceVisualPtr> covariances_;
```

Comment:
I had to copied the covariance property and visual implementation from `rviz` package because I need a container different than the one they use. In their case it's an `std::deque`, but I need to index the visual by `key` (UUID string), so I can insert and remove them efficiently.